File: NeetCode/Arrays_and_Hashing/Longest_Common_Prefix/longest__common__prefix.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        for i in range(len(strs[0])):
            for s in strs:
                if len(s) - 1 < i or s[i] != strs[0][i]:
                    return s[:i]
        return strs[0]


if __name__ == "__main__":
    sol = Solution()
    strs = ["ab", "a"]
    print(sol.longestCommonPrefix(strs))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/Longest_Common_Prefix/Longest_Common_Prefix.cpp
// Longest_Common_Prefix.cpp
//
// Compile locally with:
// clang++ Longest_Common_Prefix.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    string longestCommonPrefix(vector<string> &strs)
    {
        for (int index = 0; index < strs[0].size(); index++)
        {
            for (auto &s : strs)
            {
                if (s.size() - 1 < index || s[index] != strs[0][index])
                {
                    return s.substr(0, index);
                }
            }
        }
        return strs[0];
    }
};

int main()
{
    cout << "Running Longest_Common_Prefix.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
